### DISCUSSION TITLE
Trame shuts down automatically

### DISCUSSION THREAD

**Discussion by joserand** [Kitware/trame#379] [Q&A] [answered]
Hi, I'm running docker(vtk)example, follow readme.md, I successfully run the interface, but the connection will close  after 3-4 seconds on the browser interface, but for another docker example(no vtk ) it doesn't automatically close the connection.

When I revisit the URL, I can see the contents of the window again, but it will automatically close after a few seconds. I tried to check the logs to see if there were any errors, and the logs didn't seem to be abnormal
![image](https://github.com/Kitware/trame/assets/89247520/df26012f-2f35-432a-bddb-536e0826bfb4)

I confirmed that my code was exactly the same as the content of the example, and I deployed it according to the documentation.
Please how should I fix this error?
Thanks for anyone's reply

**Comment by jourdain** [comment]
Do you have a proxy in front of your docker?
Any browser errors in the log at disconnection time?

**Reply by joserand** [reply]
I'm not using a proxy.
When disconnecting, the browser's log content does not show any abnormality. Maybe it's just that I didn't find it,can you found the error below

    (base) [wjd@A100 VtkRendering]$ docker run --rm -it -p 8888:80 --entrypoint=/bin/bash  trame-vtk-app
    root@37a57d08cf01:/# /opt/trame/entrypoint.sh
     * Restarting Apache httpd web server apache2
     * AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.0.5. Set the 'ServerName' directive globally to suppress this message
                                                                                                                                             [ OK ]
    Starting server...
    Starting the wslink launcher at
    ======== Running on http://0.0.0.0:9000 ========
    (Press CTRL+C to quit)
    ^Croot@37a57d08cf01:/# ls
    bin  boot  deploy  dev  etc  home  lib  lib32  lib64  libx32  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
    root@37a57d08cf01:/# cd deploy
    root@37a57d08cf01:/deploy# ls
    Dockerfile  README.md  app.py  captain-definition  server  setup
    root@37a57d08cf01:/deploy# cd server
    root@37a57d08cf01:/deploy/server# ls
    launcher.json  logs  venv  www
    root@37a57d08cf01:/deploy/server# cd logs
    root@37a57d08cf01:/deploy/server/logs# ls
    apache  launcher
    root@37a57d08cf01:/deploy/server/logs# cd launcher
    root@37a57d08cf01:/deploy/server/logs/launcher# ls
    b68f1f98-8ce4-11ee-882f-0242ac110005.txt  cc175b50-8ce4-11ee-882f-0242ac110005.txt  launcherLog.log
    root@37a57d08cf01:/deploy/server/logs/launcher# cat b68f1f98-8ce4-11ee-882f-0242ac110005.txt 
    
    --------------------------------------------------------------------------------
       !!! You are currently using trame@3 which may break your application !!!
    --------------------------------------------------------------------------------
    
     1. trame@3 only provides by default trame.widgets.[html,client] and remove
        everything else as implicit dependency. Those other widgets will still
        exist and will be supported, but they will need to be defined as a
        dependency of your application.
    
           $ pip install trame-vtk trame-vuetify trame-plotly
    
        Import paths are remaining the same.
    
        For libraries like vuetify since they offer different API between
        their vue2 and vue3 implementation, the widget name will reflect
        which vue version they are referencing. But original naming will remain.
    
           from trame.widgets import vuetify2, vuetify3
    
    
     2. trame@3 aims to use vue3 as a new default. But to smooth the transition
        we will maintain the server.client_type = 'vue2' default until
        December 2023 which is the vue2 EOL.
    
        After that time, the new default will be switched to 'vue3'.
        Vue2 will still work 'forever' and many of the new widgets will be
        written to support both versions.
    
        If you have a 'vue2' application and don't need or want to update your code,
        you can still use trame@3 with vue2 by setting `server.client_type='vue2'.
    
     Actions items
     ~~~~~~~~~~~~~
       a. Make sure you set `server.client_type` to either 'vue2' or 'vue3'.
       b. List the expected dependencies or have a 'trame<3' dependency
    
    --------------------------------------------------------------------------------
     => Current client_type default: vue2
    --------------------------------------------------------------------------------
    
    
    App running at:
     - Local:   http://0.0.0.0:9500/
     - Network: http://172.17.0.5:9500/
    
    Note that for multi-users you need to use and configure a launcher.
    root@37a57d08cf01:/deploy/server/logs/launcher# cat cc175b50-8ce4-11ee-882f-0242ac110005.txt 
    
    --------------------------------------------------------------------------------
       !!! You are currently using trame@3 which may break your application !!!
    --------------------------------------------------------------------------------
    
     1. trame@3 only provides by default trame.widgets.[html,client] and remove
        everything else as implicit dependency. Those other widgets will still
        exist and will be supported, but they will need to be defined as a
        dependency of your application.
    
           $ pip install trame-vtk trame-vuetify trame-plotly
    
        Import paths are remaining the same.
    
        For libraries like vuetify since they offer different API between
        their vue2 and vue3 implementation, the widget name will reflect
        which vue version they are referencing. But original naming will remain.
    
           from trame.widgets import vuetify2, vuetify3
    
    
     2. trame@3 aims to use vue3 as a new default. But to smooth the transition
        we will maintain the server.client_type = 'vue2' default until
        December 2023 which is the vue2 EOL.
    
        After that time, the new default will be switched to 'vue3'.
        Vue2 will still work 'forever' and many of the new widgets will be
        written to support both versions.
    
        If you have a 'vue2' application and don't need or want to update your code,
        you can still use trame@3 with vue2 by setting `server.client_type='vue2'.
    
     Actions items
     ~~~~~~~~~~~~~
       a. Make sure you set `server.client_type` to either 'vue2' or 'vue3'.
       b. List the expected dependencies or have a 'trame<3' dependency
    
    --------------------------------------------------------------------------------
     => Current client_type default: vue2
    --------------------------------------------------------------------------------
    
    
    App running at:
     - Local:   http://0.0.0.0:9499/
     - Network: http://172.17.0.5:9499/
    
    Note that for multi-users you need to use and configure a launcher.
    root@37a57d08cf01:/deploy/server/logs/launcher# cat launcherLog.log 
    2023-11-27 05:35:24,520:WARNING:wslink.backends.aiohttp.launcher:Server factory shutting down. Stopping all processes
    root@37a57d08cf01:/deploy/server/logs/launcher# 

I'm sorry that this video has a low resolution because I recorded all my actions, resulting in this video being a bit long

https://github.com/Kitware/trame/assets/89247520/6586f5f6-360d-4950-a19c-93cfd53a3d4f

**Reply by jourdain** [reply, marked as answer]
It might be related to aiohttp==3.9.0 which introduced some issue. Please try to force it to be 3.8.6.

**Reply by joserand** [reply]
You're right, I used the method you provided to solve the problem of automatic shutdown very well. But I feel that this web page is more stuttering (than running app.py locally), what does this have to do with it, and what can I do to improve it?

**Reply by jourdain** [reply]
You will need to be more precise on the "stuttering" problem and maybe start another thread. In general, you should just be the same. Also `aiohttp==3.9.1` got released 2 days ago, which may have fixed that double message issue.

**Reply by joserand** [reply]
I tried what you said aiohttp=3.9.1 and yes, this version doesn't cause trame "connection close" either

**Comment by joserand** [comment]
OK，I will start another thread later. And I will try aiohttp==3.9.1.Thanks for the answer
